### PR TITLE
Train the model with AllReduce using Horovod

### LINKED ISSUE #2229: Workers use hovorod broadcast/allreduce in training and catch exceptions.

### PR DESCRIPTION
**Pull request by workingloong**
Fix #2229 and #2230

### CODE REVIEW

File: elasticdl/python/common/constants.py [hunk 1]
```diff
@@ -84,3 +84,13 @@ class ReaderType(object):
 
 class Initializer(object):
     UNIFORM = "uniform"
+
+
+class HOROVOD_CONFIG(object):
```
Comment (terrytangyuan) [inline elasticdl/python/common/constants.py]: HorovodConfig

Comment (terrytangyuan) [reply elasticdl/python/common/constants.py]: Or HorovodEnv and then remove "ENV" in the following variable names

File: elasticdl/python/worker/allreduce_trainer.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+import tensorflow as tf
+from tensorflow.python.framework.errors_impl import UnknownError
+
+from elasticdl.python.common.constants import HOROVOD_CONFIG
+
+try:
+    import horovod.tensorflow as hvd
+except ImportError:
+    pass
```
Comment (terrytangyuan) [inline elasticdl/python/worker/allreduce_trainer.py]: Why pass here?

Comment (workingloong) [reply elasticdl/python/worker/allreduce_trainer.py]: I want to catch the `ImportError` when we train the model using ParameterServerStrategy without Horovod.

File: elasticdl/python/master/master.py [hunk 3]
```diff
@@ -221,12 +223,8 @@ def prepare(self):
         if self.instance_manager:
             self.instance_manager.update_status(InstanceManagerStatus.PENDING)
             if self.distribution_strategy == DistributionStrategy.ALLREDUCE:
-                # Exposes the consensus service for allreduce-based training
-                self.rendezvous_server = HorovodRendezvousServer(
-                    os.getenv("MY_POD_IP", "localhost")
-                )
+                # Start rendezvous server for workers to initialize Horovod
                 self.rendezvous_server.start()
-                self.instance_manager.start_ftlib_consensus_service()
```
Comment (skydoorkai) [inline elasticdl/python/master/master.py:229]: Other ftlib related codes such as `start_ftlib_consensus_service` in `k8s_instance_manager.py` can also be deleted.

File: elasticdl/python/worker/worker.py [hunk 4]
```diff
@@ -202,6 +188,16 @@ def _init_from_args(self, args):
...
 
+        self._allreduce_trainer = None
+        if (
+            self._distribution_strategy == DistributionStrategy.ALLREDUCE
+            and args.num_workers > 1
+        ):
+            self._allreduce_trainer = AllReduceTrainer(
+                self._mc, self._model, self._loss, self._opt
+            )
+            self._allreduce_trainer.set_horovod_env(args.master_addr)
```
Comment (skydoorkai) [inline elasticdl/python/worker/worker.py]: master_addr can be an argument of `AllReduceTrainer` init, and `set_horovod_env` in init.

File: elasticdl/python/worker/allreduce_trainer.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        )
+        return loss
+
+    def training_process_elastic(self, features, labels):
+        for _ in range(DEFAULT_MAX_ALLREDUCE_RETRY_NUM + 1):
+            try:
+                loss = self._training_process(features, labels)
+                version = self._optimizer.iterations.numpy()
+                return version, loss
+            except UnknownError as e:
```
Comment (skydoorkai) [inline elasticdl/python/worker/allreduce_trainer.py:69]: Is it always `UnknownError`, not possible other errors?

File: elasticdl/python/worker/allreduce_trainer.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                logger.warning(
+                    "Failed to perform allreduce operation on "
+                    "the gradients. Retrying..."
+                )
+                if (
+                    "HorovodAllreduce" in e.message
+                    or "HorovodAllgather" in e.message
+                    or "HorovodBroadcast" in e.message
```
Comment (skydoorkai) [inline elasticdl/python/worker/allreduce_trainer.py:80]: These three are all possible error messages?

File: elasticdl/python/worker/allreduce_trainer.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+                loss = self._training_process(features, labels)
+                version = self._optimizer.iterations.numpy()
+                return version, loss
+            except UnknownError as e:
+                logger.warning(
+                    "Failed to perform allreduce operation on "
+                    "the gradients. Retrying..."
+                )
+                if (
+                    "HorovodAllreduce" in e.message
```
Comment (brightcoder01) [inline elasticdl/python/worker/allreduce_trainer.py:78]: Please add comments here what does these exception standards for. And in these cases, we will rebuild the communication.

File: elasticdl/python/worker/allreduce_trainer.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+                    or "HorovodAllgather" in e.message
+                    or "HorovodBroadcast" in e.message
+                ):
+                    time.sleep(3)
+                    self.init_horovod_if_needed()
+
+    def init_horovod_if_needed(self):
+        rank_response = self._master_client.get_comm_rank()
+        if rank_response.rendezvous_id != self._rendezvous_id:
```
Comment (brightcoder01) [inline elasticdl/python/worker/allreduce_trainer.py:91]: Please add comments that inequality between _rendezvous_id and rendezvous_id from master means that the communication should be rebuilt.

File: elasticdl/requirements.txt [hunk 9]
```diff
@@ -6,3 +6,4 @@ Cython
 odps
 tensorflow==2.1.0
 psutil
+cloudpickle
```
Comment (brightcoder01) [inline elasticdl/requirements.txt:9]: What is this package used for?

File: elasticdl/python/worker/allreduce_trainer.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        tape = hvd.DistributedGradientTape(tape)
+        grads = tape.gradient(loss, self._model.trainable_variables)
+        # Take care of the order of grads and vars if worker modifies
+        # `_non_embed_vars` during training.
+        self._optimizer.apply_gradients(
+            zip(grads, self._model.trainable_variables)
+        )
+        return loss
+
+    def training_process_elastic(self, features, labels):
```
Comment (brightcoder01) [inline elasticdl/python/worker/allreduce_trainer.py]: elastic_training_process or training_process_with_fault_tolerance.

I think this method focus on fault tolerance.

File: elasticdl/python/worker/allreduce_trainer.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+except ImportError:
+    hvd = None
+
+
+# The default maximum number of retries for allreduce operation
+# if allreduce-based distributed training strategy is used.
+DEFAULT_MAX_ALLREDUCE_RETRY_NUM = 5
+
+
+class AllReduceTrainer(object):
```
Comment (brightcoder01) [inline elasticdl/python/worker/allreduce_trainer.py:34]: Will we have a ParameterServerTrainer? How about move some common method into an abstract class?
These two classes will implement these methods.

File: elasticdl/python/master/master.py (hunk 3 above)
Comment (workingloong) [reply elasticdl/python/master/master.py:229]: Yes, we can delete those codes in the next PR.

File: elasticdl/python/worker/allreduce_trainer.py (hunk 6 above)
Comment (workingloong) [reply elasticdl/python/worker/allreduce_trainer.py:80]: Yes, elastic Horovod uses those three error messages.
https://github.com/horovod/horovod/blob/2bd305a4b5b247b1d78514f6e49e0969208d028c/horovod/tensorflow/elastic.py#L55-L60

File: elasticdl/requirements.txt (hunk 9 above)
Comment (workingloong) [reply elasticdl/requirements.txt:9]: Horovod needs the package.
https://github.com/horovod/horovod/blob/2bd305a4b5b247b1d78514f6e49e0969208d028c/horovod/runner/common/util/network.py#L22

File: elasticdl/python/worker/allreduce_trainer.py (hunk 10 above)
Comment (workingloong) [reply elasticdl/python/worker/allreduce_trainer.py]: yeah!

File: elasticdl/python/worker/allreduce_trainer.py (hunk 11 above)
Comment (workingloong) [reply elasticdl/python/worker/allreduce_trainer.py:34]: Yes, we can separate training codes from worker.py. I have created #2253 to develop an abstract class and ParameterServerTrainer.

File: elasticdl/python/worker/allreduce_trainer.py (hunk 5 above)
Comment (workingloong) [reply elasticdl/python/worker/allreduce_trainer.py:69]: The `UnknownError` refers to the code in elastic Horovod https://github.com/horovod/horovod/blob/2bd305a4b5b247b1d78514f6e49e0969208d028c/horovod/tensorflow/elastic.py#L55-L60

File: elasticdl/python/worker/worker.py [hunk 12]
```diff
@@ -202,6 +188,16 @@ def _init_from_args(self, args):
             checkpoint_path=args.checkpoint_dir,
         )
 
+        self._allreduce_trainer = None
+        if (
+            self._distribution_strategy == DistributionStrategy.ALLREDUCE
+            and args.num_workers > 1
```
Comment (skydoorkai) [inline elasticdl/python/worker/worker.py]: args.num_workers can be 1. Thus, remove this restriction.

Comment (workingloong) [reply elasticdl/python/worker/worker.py]: Yeah!

**Review by skydoorkai — APPROVED** [review]

**Pull request merged**